OMCompiler/Compiler/Template/CodegenFMUCommon.tpl

```diff
@@ -162,9 +162,14 @@ template ScalarVariable(SimVar simVar, SimCode simCode, list<SimVar> stateVars,
 match simVar
 case SIMVAR(type_ = T_ARRAY()) then
   /* roll out array as XML file only supports scalars */
```

Comment:
> /* roll out array as XML file only supports scalars */

This is probably slightly off-topic: @mahge Shouldn't this be done in simcode instead of the templates?

Reply:
If I understood correctly this is trying to get the elements of an array individually as simvars. In that case, it might not be even correct. The function `getScalarElements `(and `fillScalarElements `used by it) seem to assume the array elements are contiguous. This is not always the case. We have slices, missing array elements due to aliases, state vars .... In addition it only deals with the last subs which might be problematic if you have arrays of records.

There is a function (badly named) called `codegenExpSanityCheck `in _SimCodeUtil_. This seems to be the best option in my opinion. It is used to check if the elements of an array are all contiguous, if they are not it returns an ARRAY() expression of simvar elements with the correct (most of the time at least) indices. 

What you need is a function that does the expansion but even when they are contagious. You can do this by expanding the _cref_ of the array simvar using `ComponentReference.expandCref` and then calling `cref2simvar `on each _cref_ to get the simvar with the correct index.

Reply:
@lochel And, of course, yes this should be done in SimCode. At least as a call back to SimCode.